In [1]:
import pandas as pd

Загружаем таблицу с данными

In [10]:
df = pd.read_csv('DataScienceJobs.csv')

In [11]:
df # получение всей таблицы

name          address  \
0         Начинающий специалист по работе с нейросетям  Санкт-Петербург   
1                                 Junior Data Engineer  Нижний Новгород   
2                                           ML-инженер           Москва   
3                                       Data Scientist           Москва   
4                            Python Developer (Junior)            Томск   
..                                                 ...              ...   
355  Ведущий специалист по разработке скоринговых карт      Новосибирск   
356       Lead data scientist (Анализ цены и Trade in)           Москва   
357                              Senior Data Scientist            Томск   
358                            Data engineer (middle+)      Новосибирск   
359  Автор дисциплины "Задачи NLP в машинном обучении"  Нижний Новгород   

             experience                                             skills  \
0    От 1 года до 3 лет  Разработка технических заданий, Python, Сбор и...   
1    От 1 года до 3 лет                                   SQL, jvm, Hadoop   
2             Нет опыта  Знания математики, алгоритмов, структур данных...   
3    От 1 года до 3 лет  Python, Machine Learning, Математическое модел...   
4    От 1 года до 3 лет                            Python, Git, ООП, MySQL   
..                  ...                                                ...   
355  От 1 года до 3 лет  SQL, Sas, Базы данных, Скоринговая карта, База...   
356       От 3 до 6 лет  PyData: Pandas, NumPy, SciPy, scikit-learn, py...   
357         Более 6 лет  SQL, Анализ данных, A/B тесты, LTV prediction,...   
358       От 3 до 6 лет  Linux, ETL, Apache Kafka, Apache Spark, Apache...   
359       От 3 до 6 лет  NLP, ML, Data Science, Обучение и развитие, На...   

     sal_from  sal_to  
0       55000   60000  
1      150000  200000  
2      150000  170000  
3      180000  200000  
4       60000  110000  
..        ...     ...  
355     55000  105000  
356    180000  200000  
357    200000  210000  
358    150000  180000  
359    270000  320000  

[360 rows x 6 columns]

In [12]:
df.head() # получение первых пяти строк таблицы

name          address  \
0  Начинающий специалист по работе с нейросетям  Санкт-Петербург   
1                          Junior Data Engineer  Нижний Новгород   
2                                    ML-инженер           Москва   
3                                Data Scientist           Москва   
4                     Python Developer (Junior)            Томск   

           experience                                             skills  \
0  От 1 года до 3 лет  Разработка технических заданий, Python, Сбор и...   
1  От 1 года до 3 лет                                   SQL, jvm, Hadoop   
2           Нет опыта  Знания математики, алгоритмов, структур данных...   
3  От 1 года до 3 лет  Python, Machine Learning, Математическое модел...   
4  От 1 года до 3 лет                            Python, Git, ООП, MySQL   

   sal_from  sal_to  
0     55000   60000  
1    150000  200000  
2    150000  170000  
3    180000  200000  
4     60000  110000

In [13]:
df.tail() # получение последних пяти строк таблицы

name          address  \
355  Ведущий специалист по разработке скоринговых карт      Новосибирск   
356       Lead data scientist (Анализ цены и Trade in)           Москва   
357                              Senior Data Scientist            Томск   
358                            Data engineer (middle+)      Новосибирск   
359  Автор дисциплины "Задачи NLP в машинном обучении"  Нижний Новгород   

             experience                                             skills  \
355  От 1 года до 3 лет  SQL, Sas, Базы данных, Скоринговая карта, База...   
356       От 3 до 6 лет  PyData: Pandas, NumPy, SciPy, scikit-learn, py...   
357         Более 6 лет  SQL, Анализ данных, A/B тесты, LTV prediction,...   
358       От 3 до 6 лет  Linux, ETL, Apache Kafka, Apache Spark, Apache...   
359       От 3 до 6 лет  NLP, ML, Data Science, Обучение и развитие, На...   

     sal_from  sal_to  
355     55000  105000  
356    180000  200000  
357    200000  210000  
358    150000  180000  
359    270000  320000

In [14]:
df.sample(10) # получение случайных 10 строк из датафрейма

name          address  \
55                                      Data Scientist           Казань   
133                                     Data Scientist           Москва   
170             Data Scientist (Алгоритмы оптимизации)           Москва   
81                                      Data Scientist           Москва   
91                                Data Scientist (nlp)           Москва   
247                 Data Scientist (модели PD/LGD/EAD)  Нижний Новгород   
146                                      Data Engineer     Екатеринбург   
344                               DS NLP (LLM), Senior           Москва   
54   Cпециалист-исследователь в области обработки а...  Нижний Новгород   
89                        Middle/Senior Data Scientist           Москва   

             experience                                             skills  \
55        От 3 до 6 лет                                             Python   
133       От 3 до 6 лет  Уверенное владение Python. Опыт работы с библи...   
170       От 3 до 6 лет  Python, C++, Математическая статистика, Методы...   
81        От 3 до 6 лет  уверенное знание Python (библиотеки: NumPy, Pa...   
91        От 3 до 6 лет        ML, Python, CatBoost, STT, ASR, LabelStudio   
247  От 1 года до 3 лет            Python, ML, Data Science, Анализ данных   
146  От 1 года до 3 лет       Spark, Big Data, Hadoop, Базы данных, Python   
344       От 3 до 6 лет  Python, SQL, PySpark, Pandas, PyTorch, RecBole...   
54   От 1 года до 3 лет        Python, PyTorch, NLP, NER, Machine Learning   
89        От 3 до 6 лет                     Python, bash, git, SQL, Hadoop   

     sal_from  sal_to  
55      55000   85000  
133    280000  330000  
170    150000  220000  
81      90000  140000  
91     150000  180000  
247    100000  110000  
146    300000  350000  
344     85000  115000  
54     130000  160000  
89     180000  210000

Описание показателей: <br>name - название вакансии <br>address - город <br>experience - требуемый опыт <br>skills - требуемые навыки <br>sal_from - нижний порог зарплаты <br>sal_to - верхний порог зарплаты 

In [15]:
df.shape # кол-во строк и столбцов

(360, 6)

In [16]:
df.info() # сводная инофрмация о таблице

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 360 entries, 0 to 359
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        360 non-null    object
 1   address     360 non-null    object
 2   experience  360 non-null    object
 3   skills      360 non-null    object
 4   sal_from    360 non-null    int64 
 5   sal_to      360 non-null    int64 
dtypes: int64(2), object(4)
memory usage: 17.0+ KB


Таблица выше: 
1. `<class 'pandas.core.frame.DataFrame'>` – объект DataFrame
2. `RangeIndex: 360 entries, 0 to 359` – в таблице 360 строк с индексами от 0 до 359
3. `Data columns (total 6 columns):` – в таблице 6 колонок
4. После выводится небольшая таблица с информацией о названии колонки (`Column`), количестве непустых ячеек в колонке (`Non-Null Count`), а также тип данных (`Dtype`)
5. `dtypes: int64(2), object(4)` – сводка по количеству столбцов с разными типами данных
6. `memory usage: 17.0+ KB` – сколько памяти занимает таблица

Обращение к данным в таблице

`датафрейм['столбец']` или `датафрейм.столбец` – обращение к одному столбцу

`датафрейм[['столбец 1', 'столбец 2', ...]]` – обращение к нескольким столбцам

In [17]:
df['name']
# или
df.name

0           Начинающий специалист по работе с нейросетям
1                                   Junior Data Engineer
2                                             ML-инженер
3                                         Data Scientist
4                              Python Developer (Junior)
                             ...                        
355    Ведущий специалист по разработке скоринговых карт
356         Lead data scientist (Анализ цены и Trade in)
357                                Senior Data Scientist
358                              Data engineer (middle+)
359    Автор дисциплины "Задачи NLP в машинном обучении"
Name: name, Length: 360, dtype: object

In [21]:
df[['name', 'address']]

name          address
0         Начинающий специалист по работе с нейросетям  Санкт-Петербург
1                                 Junior Data Engineer  Нижний Новгород
2                                           ML-инженер           Москва
3                                       Data Scientist           Москва
4                            Python Developer (Junior)            Томск
..                                                 ...              ...
355  Ведущий специалист по разработке скоринговых карт      Новосибирск
356       Lead data scientist (Анализ цены и Trade in)           Москва
357                              Senior Data Scientist            Томск
358                            Data engineer (middle+)      Новосибирск
359  Автор дисциплины "Задачи NLP в машинном обучении"  Нижний Новгород

[360 rows x 2 columns]

Обращение к строкам

`датафрейм.loc['название строки' или индекс]` – обращение к строке по названию \
`датафрейм.iloc[индекс]` – обращение к строке по индексу (считается с 0)

In [25]:
df.loc[0]

name               Начинающий специалист по работе с нейросетям
address                                         Санкт-Петербург
experience                                   От 1 года до 3 лет
skills        Разработка технических заданий, Python, Сбор и...
sal_from                                                  55000
sal_to                                                    60000
Name: 0, dtype: object

In [26]:
df.iloc[0]

name               Начинающий специалист по работе с нейросетям
address                                         Санкт-Петербург
experience                                   От 1 года до 3 лет
skills        Разработка технических заданий, Python, Сбор и...
sal_from                                                  55000
sal_to                                                    60000
Name: 0, dtype: object

Обращение к ячейкам

`датафрейм['столбец'][индекс строки]` или `датафрейм.loc[индекс]['столбец']`

In [27]:
df['name'][0]

'Начинающий специалист по работе с нейросетям'

In [28]:
df.loc[0]['name']

'Начинающий специалист по работе с нейросетям'

In [29]:
type(df['name'][0])

str

In [30]:
df['sal_from'][0] # нижняя граница зарплаты в первой вакансии

np.int64(55000)

In [31]:
type(df['sal_from'][0])

numpy.int64

Фильтрация данных

Фильтрация по столбцу

`датафрейм[датафрейм['столбец'] <оператор сравнения> условие]`

|Оператор|Значение|
|:---:|:---:|
|==|равно|
|!=|не равно|
|>|больше чем|
|<|меньше чем|
|>=|больше или равно чем|
|<=|меньше или равно чем|

In [34]:
df[df['address'] == 'Москва']

name address  \
2                                           ML-инженер  Москва   
3                                       Data Scientist  Москва   
5                                 Data Scientist (NLP)  Москва   
6    Специалист по работе с нейронными сетями\Promt...  Москва   
7         Junior / Middle ML Engineer (Data Scientist)  Москва   
..                                                 ...     ...   
349              Руководитель направления Data Science  Москва   
351                        Data Scientist Teamlead NLP  Москва   
352  Medical foundation model research data scienti...  Москва   
354       Senior Data Scientist CV (Generative design)  Москва   
356       Lead data scientist (Анализ цены и Trade in)  Москва   

             experience                                             skills  \
2             Нет опыта  Знания математики, алгоритмов, структур данных...   
3    От 1 года до 3 лет  Python, Machine Learning, Математическое модел...   
5         От 3 до 6 лет  Python, NLP, Artificial intelligence, Искусств...   
6    От 1 года до 3 лет  Большой интерес к инновациям в области медиа и...   
7    От 1 года до 3 лет  Python, Pandas, NumPy, Transformers, openai, L...   
..                  ...                                                ...   
349       От 3 до 6 лет                                        Python, SQL   
351       От 3 до 6 лет     SQL, Python, PyTorch, Docker, Hadoop, DVC, git   
352       От 3 до 6 лет                                            PyTorch   
354       От 3 до 6 лет  Machine Learning, ML, Data Science, DS, Машинн...   
356       От 3 до 6 лет  PyData: Pandas, NumPy, SciPy, scikit-learn, py...   

     sal_from  sal_to  
2      150000  170000  
3      180000  200000  
5      200000  400000  
6      280000  330000  
7      170000  180000  
..        ...     ...  
349    280000  350000  
351    200000  270000  
352    180000  250000  
354    100000  150000  
356    180000  200000  

[188 rows x 6 columns]

In [39]:
df[df['sal_from'] >= 300000]

name             address  \
9                     Senior Machine Learning Engineer              Москва   
19                       ML-разработчик/Data Scientist     Санкт-Петербург   
35                                      Data Scientist              Москва   
42   Эксперт по моделированию и анализу данных / Da...              Москва   
49                                 ML разработчик, B2B  Московская область   
68                                      ML-разработчик        Екатеринбург   
79                                      Data Scientist               Томск   
88                                       Data Engineer         Новосибирск   
94                               Senior Data Scientist     Санкт-Петербург   
116             ML Engineer (Рекомендательные системы)         Новосибирск   
118                            Data Scientist (middle)                Сочи   
119                                     Data Scientist              Казань   
123                                    Аналитик данных         Владивосток   
132                            Middle Python Developer              Москва   
146                                      Data Engineer        Екатеринбург   
147                                         ML Инженер     Санкт-Петербург   
157                             Team-Lead Data Science              Москва   
171                                      Data Engineer              Москва   
192                              Senior Data Scientist              Москва   
204                              Senior Data Scientist        Екатеринбург   
207                           ML&Optimization engineer              Москва   
208                    Middle Computer Vision Engineer              Москва   
213     Специалист в группу технического сопровождения              Москва   
220                                      Data Engineer  Московская область   
238                                 ML Engineer RecSyS     Санкт-Петербург   
254                                 Lead Data Engineer              Москва   
265                           Data Scientist (Scoring)              Москва   
281                                        ML Engineer              Москва   
282              Системный архитектор (Data Warehouse)         Новосибирск   
294                                     Data Scientist              Москва   
303             Руководитель группы машинного обучения              Москва   
305              Стажер - разработчик скоринговых карт     Санкт-Петербург   
320               Senior MLE/DS в команду рекомендаций              Москва   
326                Ведущий Data Engineer / Lead (Java)     Санкт-Петербург   
333   Senior Data science expert в Центр технологий ИИ         Новосибирск   

             experience                                             skills  \
9         От 3 до 6 лет  Machine Learning, Python, numpy, pandas, matpl...   
19   От 1 года до 3 лет        Python, Linux, NLP, ML, TensorFlow, PyTorch   
35   От 1 года до 3 лет                       ML, Python, Machine Learning   
42   От 1 года до 3 лет     SQL, Git, Python, Анализ данных, Моделирование   
49        От 3 до 6 лет                       Python, Pyspark, Финансы, IT   
68        От 3 до 6 лет  Python, Machine Learning, RecSys, C++, ML, Dat...   
79        От 3 до 6 лет  Spark, Hive, MLFlow, ML, AirFlow, MLOps, Forec...   
88        От 3 до 6 лет  SQL, Python, ETL, ELT, СУБД, Airflow, dbt, Cli...   
94        От 3 до 6 лет  SQL, Python, Pandas, ML, machine learning, обу...   
116       От 3 до 6 лет       RecSys, рекомендательная система, ml, Python   
118  От 1 года до 3 лет                                   Python, SQL, API   
119       От 3 до 6 лет  Опыт Python разработчика (pytorch, opencv), зн...   
123  От 1 года до 3 лет  Анализ данных, Умение работать в коллективе, P...   
132  От 1 года до 3 лет  Уверенное владение R или Python (pandas/numpy/...   
146  От 1 года до 3 лет       Spark, Big Data, Hadoop, Базы данных, Pyt

In [40]:
df[df['sal_to'] == 500000]

name address     experience  \
320  Senior MLE/DS в команду рекомендаций  Москва  От 3 до 6 лет   

                    skills  sal_from  sal_to  
320  ML (GBDT, LR), Kotlin    300000  500000

Фильтрация по нескольким условиям

`&` - и \
`|` - или 

Вакансии в Москве или Казани

In [42]:
df[(df['address'] == 'Москва') | (df['address'] == 'Казань')]

name address  \
2                                           ML-инженер  Москва   
3                                       Data Scientist  Москва   
5                                 Data Scientist (NLP)  Москва   
6    Специалист по работе с нейронными сетями\Promt...  Москва   
7         Junior / Middle ML Engineer (Data Scientist)  Москва   
..                                                 ...     ...   
349              Руководитель направления Data Science  Москва   
351                        Data Scientist Teamlead NLP  Москва   
352  Medical foundation model research data scienti...  Москва   
354       Senior Data Scientist CV (Generative design)  Москва   
356       Lead data scientist (Анализ цены и Trade in)  Москва   

             experience                                             skills  \
2             Нет опыта  Знания математики, алгоритмов, структур данных...   
3    От 1 года до 3 лет  Python, Machine Learning, Математическое модел...   
5         От 3 до 6 лет  Python, NLP, Artificial intelligence, Искусств...   
6    От 1 года до 3 лет  Большой интерес к инновациям в области медиа и...   
7    От 1 года до 3 лет  Python, Pandas, NumPy, Transformers, openai, L...   
..                  ...                                                ...   
349       От 3 до 6 лет                                        Python, SQL   
351       От 3 до 6 лет     SQL, Python, PyTorch, Docker, Hadoop, DVC, git   
352       От 3 до 6 лет                                            PyTorch   
354       От 3 до 6 лет  Machine Learning, ML, Data Science, DS, Машинн...   
356       От 3 до 6 лет  PyData: Pandas, NumPy, SciPy, scikit-learn, py...   

     sal_from  sal_to  
2      150000  170000  
3      180000  200000  
5      200000  400000  
6      280000  330000  
7      170000  180000  
..        ...     ...  
349    280000  350000  
351    200000  270000  
352    180000  250000  
354    100000  150000  
356    180000  200000  

[203 rows x 6 columns]

Вакансии в Москве, без опыта работы

In [44]:
df[(df['address'] == 'Москва') & (df['experience'] == 'Нет опыта')]

name address experience  \
2                                 ML-инженер  Москва  Нет опыта   
15                         Стажер ML-инженер  Москва  Нет опыта   
17                       Стажер Data analyst  Москва  Нет опыта   
20                    Стажёр - Дата аналитик  Москва  Нет опыта   
34                Solution engineer (Junior)  Москва  Нет опыта   
39                   Стажёр - Data Scientist  Москва  Нет опыта   
72                       Lead Data Scientist  Москва  Нет опыта   
83  Machine learning engineer, Middle/Senior  Москва  Нет опыта   

                                               skills  sal_from  sal_to  
2   Знания математики, алгоритмов, структур данных...    150000  170000  
15  Знание основ SQL, Базовое знание Python, Знани...    280000  300000  
17  Писать SQL-запросы (у нас Greenplum), Анализир...     55000   55000  
20  Обучение и развитие, Аналитика, Системное мышл...    150000  180000  
34                                  Python, SQL, bash    100000  120000  
39  Обучение и развитие, Аналитика, Системное мышл...    200000  250000  
72                                                 ML    180000  230000  
83                           Linux, Git, Bash, Docker    180000  190000

Вакансии в Москве или Казани, с опытом от 3 до 6 лет и зарплатой от 200000

In [55]:
df[((df['address'] == 'Москва') | (df['address'] == 'Казань')) 
   & (df['experience'] == 'От 3 до 6 лет') 
   & (df['sal_from'] >= 200000)]

name address     experience  \
5                                 Data Scientist (NLP)  Москва  От 3 до 6 лет   
9                     Senior Machine Learning Engineer  Москва  От 3 до 6 лет   
10                             Computer vision инженер  Москва  От 3 до 6 лет   
11                                      Data Scientist  Москва  От 3 до 6 лет   
16                                       ML специалист  Москва  От 3 до 6 лет   
25                           Machine learning engineer  Москва  От 3 до 6 лет   
58                                Аналитик-разработчик  Москва  От 3 до 6 лет   
71                        Data Scientist middle/senior  Москва  От 3 до 6 лет   
86                                     Аналитик данных  Москва  От 3 до 6 лет   
108                                Senior Data Science  Москва  От 3 до 6 лет   
111                   Дата-инженер (Motion Prediction)  Москва  От 3 до 6 лет   
119                                     Data Scientist  Казань  От 3 до 6 лет   
128                  Data Scientist в команду аукциона  Москва  От 3 до 6 лет   
129                                  Архитектор данных  Москва  От 3 до 6 лет   
133                                     Data Scientist  Москва  От 3 до 6 лет   
152                          Machine Learning Engineer  Москва  От 3 до 6 лет   
157                             Team-Lead Data Science  Москва  От 3 до 6 лет   
167                                     Data Scientist  Москва  От 3 до 6 лет   
196                Data Scientist, Ранжирование поиска  Москва  От 3 до 6 лет   
207                           ML&Optimization engineer  Москва  От 3 до 6 лет   
210                         Data Engineer (Perception)  Москва  От 3 до 6 лет   
212                Data scientist TTS (Text To Speech)  Москва  От 3 до 6 лет   
237                                    Разработчик ETL  Москва  От 3 до 6 лет   
240        Machine Learning / Computer Vision Engineer  Москва  От 3 до 6 лет   
244                         Senior/Lead Data Scientist  Москва  От 3 до 6 лет   
254                                 Lead Data Engineer  Москва  От 3 до 6 лет   
255                                     Data Scientist  Москва  От 3 до 6 лет   
256                              Teamlead Data Science  Москва  От 3 до 6 лет   
257                              Senior Data scientist  Москва  От 3 до 6 лет   
263                             Data Scienti​​​​​​​st   Москва  От 3 до 6 лет   
265                           Data Scientist (Scoring)  Москва  От 3 до 6 лет   
270                 ML Data Scientist (Middle, Senior)  Москва  От 3 до 6 лет   
274                                     Data scientist  Москва  От 3 до 6 лет   
277  Senior Data инженер в команду Data Финансовой ...  Москва  От 3 до 6 лет   
280                                 Lead Data Engineer  Москва  От 3 до 6 лет   
281                                        ML Engineer  Москва  От 3 до 6 лет   
287               Computer Vision engineer (Team lead)  Москва  От 3 до 6 лет   
294                                     Data Scientist  Москва  От 3 до 6 лет   
302                  Team Lead Data Scientist (RecSys)  Казань  От 3 до 6 лет   
303             Руководитель группы машинного обучения  Москва  От 3 до 6 лет   
307                                     Data Scientist  Москва  От 3 до 6 лет   
320               Senior MLE/DS в команду рекомендаций  Москва  От 3 до 6 лет   
323                    Middle NLP Data Scientist (LLM)  Москва  От 3 до 6 лет   
328                                Senior NLP Engineer  Москва  От 3 до 6 лет   
332                                       NLP engineer  Москва  От 3 до 6 лет   
338                    Middle NLP Data Scientist (LLM)  Москва  От 3 до 6 лет   
342                     Machine Learning Lead engineer  Москва  От 3 до 6 лет   
349              Руководитель направления Data Science  Москва  От 3 до 6 лет   
351                        Data Scientist Teamlead NLP  Москва  От 3 до 6 лет   


Подсчет строк после фильтрации через .shape

In [56]:
df[((df['address'] == 'Москва') | (df['address'] == 'Казань')) 
   & (df['experience'] == 'От 3 до 6 лет') 
   & (df['sal_from'] >= 200000)].shape

(49, 6)

Сортировка данных

По одному условию

`датафрейм.sort_values('столбец')` – сортировка по одному столбцу (от наименьшего к наибольшему)
<br>`датафрейм.sort_values('столбец', ascending=False)` – сортировка по одному столбцу (от наибольшего к наименьшему)

In [59]:
df.sort_values('sal_from')

name          address  \
275     Специалист в группу технического сопровождения      Новосибирск   
0         Начинающий специалист по работе с нейросетям  Санкт-Петербург   
289                              Senior NLP Researcher           Казань   
217  Data Scientist (Deep Learning for Demand Forec...  Санкт-Петербург   
55                                      Data Scientist           Казань   
..                                                 ...              ...   
119                                     Data Scientist           Казань   
118                            Data Scientist (middle)             Сочи   
192                              Senior Data Scientist           Москва   
204                              Senior Data Scientist     Екатеринбург   
9                     Senior Machine Learning Engineer           Москва   

             experience                                             skills  \
275  От 1 года до 3 лет  MS SQL, PostgreSQL, API, Базы данных, Техничес...   
0    От 1 года до 3 лет  Разработка технических заданий, Python, Сбор и...   
289       От 3 до 6 лет  PyTorch, TensorFlow, Transformers, T5, Bert, H...   
217       От 3 до 6 лет        Deep Learning, Прогнозирование, Time Series   
55        От 3 до 6 лет                                             Python   
..                  ...                                                ...   
119       От 3 до 6 лет  Опыт Python разработчика (pytorch, opencv), зн...   
118  От 1 года до 3 лет                                   Python, SQL, API   
192  От 1 года до 3 лет  SQL, MS SQL, Big Data, Python, Spark, Hadoop, ...   
204       От 3 до 6 лет  Python, SQL, Spark, Базы данных, PHP, ClickHou...   
9         От 3 до 6 лет  Machine Learning, Python, numpy, pandas, matpl...   

     sal_from  sal_to  
275     45000  145000  
0       55000   60000  
289     55000  125000  
217     55000  105000  
55      55000   85000  
..        ...     ...  
119    300000  330000  
118    300000  370000  
192    300000  350000  
204    330000  450000  
9      350000  570000  

[360 rows x 6 columns]

In [60]:
df.sort_values('sal_from', ascending=False)

name             address  \
9                     Senior Machine Learning Engineer              Москва   
204                              Senior Data Scientist        Екатеринбург   
49                                 ML разработчик, B2B  Московская область   
119                                     Data Scientist              Казань   
305              Стажер - разработчик скоринговых карт     Санкт-Петербург   
..                                                 ...                 ...   
60                                      Data Scientist              Тюмень   
186                                      Data Engineer         Новосибирск   
217  Data Scientist (Deep Learning for Demand Forec...     Санкт-Петербург   
0         Начинающий специалист по работе с нейросетям     Санкт-Петербург   
275     Специалист в группу технического сопровождения         Новосибирск   

             experience                                             skills  \
9         От 3 до 6 лет  Machine Learning, Python, numpy, pandas, matpl...   
204       От 3 до 6 лет  Python, SQL, Spark, Базы данных, PHP, ClickHou...   
49        От 3 до 6 лет                       Python, Pyspark, Финансы, IT   
119       От 3 до 6 лет  Опыт Python разработчика (pytorch, opencv), зн...   
305           Нет опыта  SQL, Python, Базы данных, Скоринговая карта, Б...   
..                  ...                                                ...   
60   От 1 года до 3 лет               sklearn, scipy, pytorch, Git, Docker   
186       От 3 до 6 лет  SQL, Python, Airflow, ORACLE, Postgres, Git, C...   
217       От 3 до 6 лет        Deep Learning, Прогнозирование, Time Series   
0    От 1 года до 3 лет  Разработка технических заданий, Python, Сбор и...   
275  От 1 года до 3 лет  MS SQL, PostgreSQL, API, Базы данных, Техничес...   

     sal_from  sal_to  
9      350000  570000  
204    330000  450000  
49     300000  320000  
119    300000  330000  
305    300000  370000  
..        ...     ...  
60      55000  105000  
186     55000   65000  
217     55000  105000  
0       55000   60000  
275     45000  145000  

[360 rows x 6 columns]

Сортировка по нескольким условиям

`датафрейм.sort_values(['столбец1', 'столбец2', ...])` – сортировка по нескольким столбцам от наименьшего к наибольшему \
`датафрейм.sort_values(['столбец1', 'столбец2', ...], ascending=False)` – сортировка по нескольким столбцам от наибольшего к наименьшему \
`датафрейм.sort_values(['столбец1', 'столбец2'], ascending=[False, True])` – сортировка сначала по первому столбцу от наибольшего к наименьшему, затем по второму столбцу от наименьшего к наибольшему <br>Сортировка слева направо

In [61]:
df.sort_values(['sal_from', 'sal_to'])

name          address  \
275    Специалист в группу технического сопровождения      Новосибирск   
17                                Стажер Data analyst           Москва   
0        Начинающий специалист по работе с нейросетям  Санкт-Петербург   
186                                     Data Engineer      Новосибирск   
55                                     Data Scientist           Казань   
..                                                ...              ...   
305             Стажер - разработчик скоринговых карт  Санкт-Петербург   
333  Senior Data science expert в Центр технологий ИИ      Новосибирск   
320              Senior MLE/DS в команду рекомендаций           Москва   
204                             Senior Data Scientist     Екатеринбург   
9                    Senior Machine Learning Engineer           Москва   

             experience                                             skills  \
275  От 1 года до 3 лет  MS SQL, PostgreSQL, API, Базы данных, Техничес...   
17            Нет опыта  Писать SQL-запросы (у нас Greenplum), Анализир...   
0    От 1 года до 3 лет  Разработка технических заданий, Python, Сбор и...   
186       От 3 до 6 лет  SQL, Python, Airflow, ORACLE, Postgres, Git, C...   
55        От 3 до 6 лет                                             Python   
..                  ...                                                ...   
305           Нет опыта  SQL, Python, Базы данных, Скоринговая карта, Б...   
333       От 3 до 6 лет  Python, SQL, алгоритмы ML, Pandas, Sklearn, Nu...   
320       От 3 до 6 лет                              ML (GBDT, LR), Kotlin   
204       От 3 до 6 лет  Python, SQL, Spark, Базы данных, PHP, ClickHou...   
9         От 3 до 6 лет  Machine Learning, Python, numpy, pandas, matpl...   

     sal_from  sal_to  
275     45000  145000  
17      55000   55000  
0       55000   60000  
186     55000   65000  
55      55000   85000  
..        ...     ...  
305    300000  370000  
333    300000  400000  
320    300000  500000  
204    330000  450000  
9      350000  570000  

[360 rows x 6 columns]

In [62]:
df.sort_values(['sal_from', 'sal_to'], ascending=False)

name          address  \
9                    Senior Machine Learning Engineer           Москва   
204                             Senior Data Scientist     Екатеринбург   
320              Senior MLE/DS в команду рекомендаций           Москва   
333  Senior Data science expert в Центр технологий ИИ      Новосибирск   
79                                     Data Scientist            Томск   
..                                                ...              ...   
55                                     Data Scientist           Казань   
186                                     Data Engineer      Новосибирск   
0        Начинающий специалист по работе с нейросетям  Санкт-Петербург   
17                                Стажер Data analyst           Москва   
275    Специалист в группу технического сопровождения      Новосибирск   

             experience                                             skills  \
9         От 3 до 6 лет  Machine Learning, Python, numpy, pandas, matpl...   
204       От 3 до 6 лет  Python, SQL, Spark, Базы данных, PHP, ClickHou...   
320       От 3 до 6 лет                              ML (GBDT, LR), Kotlin   
333       От 3 до 6 лет  Python, SQL, алгоритмы ML, Pandas, Sklearn, Nu...   
79        От 3 до 6 лет  Spark, Hive, MLFlow, ML, AirFlow, MLOps, Forec...   
..                  ...                                                ...   
55        От 3 до 6 лет                                             Python   
186       От 3 до 6 лет  SQL, Python, Airflow, ORACLE, Postgres, Git, C...   
0    От 1 года до 3 лет  Разработка технических заданий, Python, Сбор и...   
17            Нет опыта  Писать SQL-запросы (у нас Greenplum), Анализир...   
275  От 1 года до 3 лет  MS SQL, PostgreSQL, API, Базы данных, Техничес...   

     sal_from  sal_to  
9      350000  570000  
204    330000  450000  
320    300000  500000  
333    300000  400000  
79     300000  370000  
..        ...     ...  
55      55000   85000  
186     55000   65000  
0       55000   60000  
17      55000   55000  
275     45000  145000  

[360 rows x 6 columns]

In [63]:
df.sort_values(['sal_from', 'sal_to'], ascending=[False, True])

name          address  \
9                     Senior Machine Learning Engineer           Москва   
204                              Senior Data Scientist     Екатеринбург   
42   Эксперт по моделированию и анализу данных / Da...           Москва   
171                                      Data Engineer           Москва   
208                    Middle Computer Vision Engineer           Москва   
..                                                 ...              ...   
60                                      Data Scientist           Тюмень   
217  Data Scientist (Deep Learning for Demand Forec...  Санкт-Петербург   
355  Ведущий специалист по разработке скоринговых карт      Новосибирск   
289                              Senior NLP Researcher           Казань   
275     Специалист в группу технического сопровождения      Новосибирск   

             experience                                             skills  \
9         От 3 до 6 лет  Machine Learning, Python, numpy, pandas, matpl...   
204       От 3 до 6 лет  Python, SQL, Spark, Базы данных, PHP, ClickHou...   
42   От 1 года до 3 лет     SQL, Git, Python, Анализ данных, Моделирование   
171  От 1 года до 3 лет  Scala / Python Hadoop (hdfs, hive) Apache Spar...   
208  От 1 года до 3 лет  NumPy, Pandas, OpenCV, Sklearn, Matplotlib, CI/CD   
..                  ...                                                ...   
60   От 1 года до 3 лет               sklearn, scipy, pytorch, Git, Docker   
217       От 3 до 6 лет        Deep Learning, Прогнозирование, Time Series   
355  От 1 года до 3 лет  SQL, Sas, Базы данных, Скоринговая карта, База...   
289       От 3 до 6 лет  PyTorch, TensorFlow, Transformers, T5, Bert, H...   
275  От 1 года до 3 лет  MS SQL, PostgreSQL, API, Базы данных, Техничес...   

     sal_from  sal_to  
9      350000  570000  
204    330000  450000  
42     300000  310000  
171    300000  310000  
208    300000  310000  
..        ...     ...  
60      55000  105000  
217     55000  105000  
355     55000  105000  
289     55000  125000  
275     45000  145000  

[360 rows x 6 columns]